File: DSA/DSAPatterns/MultiplePointers/JS/SmallestUnsortedSubarray.js
// https://leetcode.com/problems/shortest-unsorted-continuous-subarray/

function findSmallestUnsortedSubarray(nums) {
	let leftPtr = 0;
	let rightPtr = nums.length - 1;
	let prevLeft;
	let prevRight;

	while (leftPtr < rightPtr) {
		// increament the left pointer till we find the element that is greater than the next element in the array.
		if (leftPtr + 1 < nums.length && nums[leftPtr] <= nums[leftPtr + 1]) {
			leftPtr++;
		}
		// decrement the right pointer till we find the element that is smaller than the previous element in the array.
		if (rightPtr > 0 && nums[rightPtr] >= nums[rightPtr - 1]) {
			rightPtr--;
		}

		// if previous leftPtr and prev rightPtr are same as current then break the loop.
		if (prevLeft == leftPtr && prevRight == rightPtr) {
			break;
		}

		// updating the prevLeft and prevRight with current one for the next iteration of loop
		prevLeft = leftPtr;
		prevRight = rightPtr;
	}

	// If leftPtr === rightPtr, then it means that our array is already sorted.
	if (leftPtr === rightPtr) {
		return 0;
	}

	const subArr = nums.slice(leftPtr, rightPtr + 1);
	const subArrMin = Math.min(...subArr);
	const subArrMax = Math.max(...subArr);

	// extending subArr window to the left to include elements that are greater than the subArrMin;
	while (leftPtr > 0 && nums[leftPtr - 1] > subArrMin) {
		leftPtr--;
	}
	// extending subArr window to the right to include elements that are lesser than the subArrMax;
	while (rightPtr + 1 < nums.length && nums[rightPtr + 1] < subArrMax) {
		rightPtr++;
	}

	return rightPtr - leftPtr + 1;
}

// TC - O(n)
function main() {
	console.log(findSmallestUnsortedSubarray([1, 3, 2, 0, -1, 7, 10])); // 5
	console.log(findSmallestUnsortedSubarray([1, 2, 5, 7, 3, 10, 11, 12])); // 3
	console.log(findSmallestUnsortedSubarray([1, 2, 3, 3, 3])); // 0
	console.log(findSmallestUnsortedSubarray([4, 3, 2, 1])); // 4
	console.log(findSmallestUnsortedSubarray([12, 7, 8, 1, 2, 0, 10, 11])); // 8
}

main();

File: DSA/DSAPatterns/MultiplePointers/C++/SmallestUnsortedSubarray.cpp
// https://leetcode.com/problems/shortest-unsorted-continuous-subarray/]

#include <bits/stdc++.h>
using namespace std;

int findSmallestUnsortedSubarray(vector<int> nums) {
  // O(n) solution
  int n = nums.size();
  int end = -1, start = -1;
  int maxi = nums[0], mini = nums[n - 1];

  for (int i = 1; i < n; i++) {
    // storing the max of the array till index i
    maxi = max(maxi, nums[i]);
    // if the current max is greater than the current element then change the
    // end ptr to i;
    if (nums[i] < maxi)
      end = i;
  }
  for (int i = n - 2; i >= 0; i--) {
    // storing the min of the array till index i
    mini = min(mini, nums[i]);
    // if the current min is lesser than the current element then change the
    // start ptr to i;
    if (nums[i] > mini)
      start = i;
  }

  // if either of the end or start pointer is -1 then it means array is
  // sorted,so return 0
  if (end == -1 || start == -1)
    return 0;

  // return length of the unsorted subarray
  return end - start + 1;
}

// TC - O(n)
int main() {
  cout << findSmallestUnsortedSubarray({1, 3, 2, 0, -1, 7, 10}) << endl;
  cout << findSmallestUnsortedSubarray({1, 2, 5, 7, 3, 10, 11, 12}) << endl;
  cout << findSmallestUnsortedSubarray({1, 2, 3, 3, 3}) << endl;
  cout << findSmallestUnsortedSubarray({4, 3, 2, 1}) << endl;
  cout << findSmallestUnsortedSubarray({12, 7, 8, 1, 2, 0, 10, 11}) << endl;
  return 0;
}